BaragonAgentService/src/main/java/com/hubspot/baragon/agent/managers/AgentRequestManager.java

```diff
@@ -97,32 +132,47 @@ private AgentBatchResponseItem getResponseItem(Response httpResponse, BaragonReq
+  public Response processRequest(String requestId, Optional<BaragonRequest> maybeRequest, Map<String, Collection<UpstreamInfo>> existingUpstreams, Optional<RequestAction> maybeAction, boolean delayReload, Optional<Integer> batchItemNumber) throws InterruptedException {
     if (!maybeRequest.isPresent()) {
       return Response.status(Response.Status.NOT_FOUND).entity(String.format("Request %s does not exist", requestId)).build();
     }
     final BaragonRequest request = maybeRequest.get();
     RequestAction action = maybeAction.or(request.getAction().or(RequestAction.UPDATE));
     Optional<BaragonService> maybeOldService = getOldService(request);
 
-    return processRequest(requestId, action, request, maybeOldService, delayReload, batchItemNumber);
+    return processRequest(requestId, action, request, maybeOldService, existingUpstreams, delayReload, batchItemNumber);
   }
 
-  public Response processRequest(String requestId, RequestAction action, BaragonRequest request, Optional<BaragonService> maybeOldService, boolean delayReload, Optional<Integer> batchItemNumber) {
+  public Response processRequest(String requestId, RequestAction action, BaragonRequest request, Optional<BaragonService> maybeOldService, Map<String, Collection<UpstreamInfo>> existingUpstreams, boolean delayReload, Optional<Integer> batchItemNumber) {
     long start = System.currentTimeMillis();
     try {
       agentState.set(BaragonAgentState.APPLYING);
       LOG.info("Received request to {} with id {}", action, requestId);
+      Collection<UpstreamInfo> existingUpstreamsForThisService;
+      String serviceId;
       switch (action) {
         case DELETE:
           return delete(request, maybeOldService, delayReload);
         case RELOAD:
           return reload(request, delayReload);
         case REVERT:
-          return revert(request, maybeOldService, delayReload, batchItemNumber);
+          serviceId = maybeOldService.or(request.getLoadBalancerService()).getServiceId();
+          existingUpstreamsForThisService = existingUpstreams.get(serviceId);
+          if (existingUpstreamsForThisService == null || existingUpstreamsForThisService.isEmpty()) {
+            existingUpstreamsForThisService = new ArrayList<>(stateDatastore.getUpstreams(serviceId));
+            existingUpstreams.put(serviceId, existingUpstreamsForThisService);
+          }
+
+          return revert(request, maybeOldService, existingUpstreamsForThisService, delayReload, batchItemNumber);
         default:
-          return apply(request, maybeOldService, delayReload, batchItemNumber);
+          serviceId = request.getLoadBalancerService().getServiceId();
+          existingUpstreamsForThisService = existingUpstreams.get(serviceId);
+          if (existingUpstreamsForThisService == null || existingUpstreamsForThisService.isEmpty()) {
+            existingUpstreamsForThisService = new ArrayList<>(stateDatastore.getUpstreams(serviceId));
```

Comment:
why are we resetting from the state datastore here? If it's empty wouldn't we assume that it's because we already looked it up and it was empty earlier?

Reply:
This is mostly to cover the case where we had an exception grabbing the upstreams the first time, for whatever reason: https://github.com/HubSpot/Baragon/pull/309/files/c8b670ab071ebc5d957307fdb5e8373560e21452#diff-2829d66da65c086b8c7d4fbb094196d3R100

Although I suppose we could just throw a RuntimeException there instead, which is what we do elsewhere in the code when try/catching around `stateDatastore.getUpstreams()`.